lib/pavilion/scriptcomposer.py

```diff
@@ -0,0 +1,697 @@
+#     and the following disclaimer in the documentation 
+#     and/or other materials provided with the distribution. 
+#  -  Neither the name of Los Alamos National Security, LLC, 
+#     Los Alamos National Laboratory, LANL, the U.S. Government, 
+#     nor the names of its contributors may be used to endorse 
+#     or promote products derived from this software without 
+#     specific prior written permission.
+#   
+#  THIS SOFTWARE IS PROVIDED BY LOS ALAMOS NATIONAL SECURITY, LLC 
+#  AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+#  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+#  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. 
+#  IN NO EVENT SHALL LOS ALAMOS NATIONAL SECURITY, LLC OR CONTRIBUTORS 
+#  BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, 
+#  OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, 
+#  OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY 
+#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
+#  TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
+#  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY 
+#  OF SUCH DAMAGE.
+#
+#  ###################################################################
+
+from __future__ import unicode_literals
+import os, datetime, grp, pwd, stat
+from collections import OrderedDict
+
+""" Class to allow for scripts to be written for other modules.
+    Typically, this will be used to write bash or batch scripts. 
+"""
+
+
+class scriptHeader( object ):
+    """Class to serve as a struct for the script header."""
+
+    @property
+    def shell_path( self ):
+        """Function to return the value of the internal shell path variable."""
+        return self._shell_path
+
+    @shell_path.setter
+    def shell_path( self, value ):
+        """Function to set the value of the internal shell path variable."""
+        if value != None and not isinstance(value, unicode):
+            error = "Shell Path must be of type 'str', not {}".format(
+                    type( value ) )
+            raise TypeError( error )
+
+        self._shell_path = value
+
+    @property
+    def scheduler_macros( self ):
+        """Function to return the value of the internal scheduler macros
+        variable.
+        """
+        return self._scheduler_macros
+
+    @scheduler_macros.setter
+    def scheduler_macros( self, value ):
+        """Function to set the value of the internal scheduler macros
+        variable.
+        """
+        if value != None and not isinstance(value, OrderedDict):
+            error = "Scheduler Macro must be of type 'OrderedDict', not " +\
+                    "{}".format( type( value ) )
+            raise TypeError( error )
+
+        self._scheduler_macros = value
+
+    def __init__( self, shell_path=None, scheduler_macros=None ):
```

Comment:
This is nitpicky, but __init__ should really be the first method listed.

Reply:
**init** is not first here intentionally.  By using the setters inside of the **init** function, it uses the type checkers.  I'm operating under the assumption that the setters will need to be defined before they are used.  If that's not true, I can definitely move the **init** function up on every class.

Reply:
Definition order only matters inside at each level of scoping. The contents of __init__ don't care about whether the getters were defined before or after __init__, as they'll only be executed after the scope resolves in full.